client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -157,22 +162,37 @@ private void handleMissingLog() {
      * message find which new segment it should go to and send it there. 
      */
     private void handleLogSealed(Segment segment) {
-        retransmitPool.submit(() -> {
-            writeFlushLock.readLock().lock();
-            /* Using segmentSealedLock the following behaviour is enforced
-               - Prevent concurrent segmentSealedCallback for different segments from being invoked concurrently.
-               - Ensure waiting segmentSealedCallbacks are invoked before the next write is invoked.
-               This ensures that resend() would be invoked again if we observe a segment sealed exception.
-             */
-            segmentSealedLock.lock();
-            try {
-                List<PendingEvent> toResend = selector.refreshSegmentEventWritersUponSealed(segment, segmentSealedCallBack);
-                resend(toResend);
-            } finally {
-                segmentSealedLock.unlock();
-                writeFlushLock.readLock().unlock();
-            }
-        });
+        /*
+         * Using segmentSealedLock prevents concurrent segmentSealedCallback for different segments
+         * from being invoked concurrently.
+         * 
+         * By calling flush when while the write lock is held we can ensure that any inflight
+         * entries that will succeed in being written to a new segment are written and any
+         * segmentSealedCallbacks will be be called happen before the next write is invoked.
+         */
+        writeFlushLock.readLock().lock();
```

Comment:
One thing that worries me deeply here is that we acquire the lock synchronously, thus blocking the calling thread, but the operation is done asynchronously. This can quickly lead to thread exhaustion. Since this is an async operation, I strongly suggest implementing/using an async lock.

I think at one point I did write an Async Lock somewhere in common but I deleted it since it was no longer necessary.